Code/FindRGL.cmake

```diff
@@ -11,78 +11,100 @@
-            PATTERNS ${API_REL_PATH}/*
-            VERBOSE
-            )
-
-    # Move the extracted files to their desired locations
-    file(RENAME ${DEST_SO_DIR}/${SO_REL_PATH} ${DEST_SO_DIR}/${SO_FILENAME})
-    file(COPY ${DEST_API_DIR}/${API_REL_PATH} DESTINATION ${DEST_SO_DIR}/include/rgl/)
+set(RGL_DOWNLOAD_IN_PROGRESS_FILE ${CMAKE_CURRENT_BINARY_DIR}/RGL_DOWNLOAD_IN_PROGRESS)
+if (NOT EXISTS ${RGL_DOWNLOAD_IN_PROGRESS_FILE})
+    FILE(TOUCH ${RGL_DOWNLOAD_IN_PROGRESS_FILE})
+
+    # Read metadata
+    set(RGL_VERSION_METADATA " ")
+    set(ROS_DISTRO_METADATA " ")
+    if (EXISTS ${RGL_VERSION_METADATA_FILE})
+        file(READ ${RGL_VERSION_METADATA_FILE} RGL_VERSION_METADATA)
+    endif ()
+    if (EXISTS ${ROS_DISTRO_METADATA_FILE})
+        file(READ ${ROS_DISTRO_METADATA_FILE} ROS_DISTRO_METADATA)
+    endif ()
+
+    # If metadata does not match, download RGL
+    if ((NOT ${RGL_VERSION_METADATA} STREQUAL ${RGL_VERSION}) OR (NOT ${ROS_DISTRO_METADATA} STREQUAL ${ROS_DISTRO}))
+        message("Downloading RGL " ${RGL_VERSION} " for ROS " ${ROS_DISTRO} "...")
+
+        # Download the RGL archive files
+        file(DOWNLOAD
+                ${RGL_LINUX_ZIP_URL}
+                ${DEST_SO_DIR}/${RGL_LINUX_ZIP_FILENAME}
+        )
+
+        # Extract the contents of the downloaded archive files
+        file(ARCHIVE_EXTRACT INPUT ${DEST_SO_DIR}/${RGL_LINUX_ZIP_FILENAME}
+                DESTINATION ${DEST_SO_DIR}
+                PATTERNS ${SO_REL_PATH}
+                VERBOSE
+        )
+
+        # Move the extracted files to their desired locations
+        file(RENAME ${DEST_SO_DIR}/${SO_REL_PATH} ${DEST_SO_DIR}/${SO_FILENAME})
+
+        # Remove the unwanted byproducts
+        file(REMOVE ${DEST_SO_DIR}/${RGL_LINUX_ZIP_FILENAME})
+
+        # Download API headers
+        file(DOWNLOAD
+                ${RGL_SRC_ROOT_URL}/include/rgl/api/core.h
+                ${DEST_API_DIR}/core.h
+        )
+        file(DOWNLOAD
+                ${RGL_SRC_ROOT_URL}/extensions/ros2/include/rgl/api/extensions/ros2.h
+                ${DEST_API_DIR}/extensions/ros2.h
+        )
+
+        # Save current metadata
+        file(WRITE ${RGL_VERSION_METADATA_FILE} ${RGL_VERSION})
+        file(WRITE ${ROS_DISTRO_METADATA_FILE} ${ROS_DISTRO})
+    endif ()
 
     # Remove the unwanted byproducts
-    file(REMOVE_RECURSE ${DEST_API_DIR}/${RGL_SRC_ZIP_FILENAME_BASE})
-    file(REMOVE ${DEST_API_DIR}/${RGL_SRC_ZIP_FILENAME})
-    file(REMOVE ${DEST_SO_DIR}/${RGL_LINUX_ZIP_FILENAME})
-
-    file(REMOVE ${CMAKE_CURRENT_SOURCE_DIR}/DOWNLOAD_RGL)
+    file(REMOVE ${RGL_DOWNLOAD_IN_PROGRESS_FILE})
 else ()
     message(WARNING "Omitting the RobotecGPULidar library download. This is intended when using the Clion multi-profile"
-    " project reload. This may also happen due to interruption of previous project configurations. If you have"
-    " any issues related to the libRobotecGPULidar.so file please make sure that the DOWNLOAD_RGL file is not"
```

Comment:
I would suggest a more specific instruction as clearing cmake cache could be a costly operation.